File: src/agentshell/commands/tree.py
from __future__ import annotations

import os
from pathlib import Path

from agentshell.fs import IGNORED_NAMES, TREE_MAX_DEPTH, TREE_MAX_FILES
from agentshell.output import CommandResult, csv, join_lines


IMPORTANT_ORDER = [
    "pyproject.toml",
    "package.json",
    "Cargo.toml",
    "go.mod",
    "README.md",
    "src",
    "tests",
]


def run(path: str = ".") -> CommandResult:
    root = Path(path)
    files_seen = 0
    dirs_seen = 0
    omitted: set[str] = set()
    groups: dict[str, list[str]] = {}
    truncated = False

    project = _detect_project(root)
    important = _important(root)

    for current, dirnames, filenames in os.walk(root):
        current_path = Path(current)
        rel = current_path.relative_to(root)
        depth = 0 if rel == Path(".") else len(rel.parts)
        if depth >= TREE_MAX_DEPTH:
            omitted.update(dirnames)
            dirnames[:] = []
        else:
            kept = []
            for name in sorted(dirnames):
                child = current_path / name
                if name in IGNORED_NAMES:
                    omitted.add(name)
                elif child.is_symlink():
                    omitted.add(f"{name}@")
                else:
                    kept.append(name)
                    dirs_seen += 1
            dirnames[:] = kept

        visible_files = []
        for name in sorted(filenames):
            if files_seen >= TREE_MAX_FILES:
                truncated = True
                break
            if name in IGNORED_NAMES:
                omitted.add(name)
                continue
            visible_files.append(name)
            files_seen += 1
        key = "root" if rel == Path(".") else str(rel)
        entries = _ordered(visible_files + dirnames)
        if entries:
            groups[key] = entries[:12]
        if truncated:
            dirnames[:] = []
            break

    lines = [
        f"project={project} files={files_seen} dirs={dirs_seen} important={csv(important)}",
    ]
    lines.extend(f"{name}: {csv(entries)}" for name, entries in sorted(groups.items(), key=_group_sort))
    if omitted or truncated:
        values = sorted(omitted)
        if truncated:
            values.append("max_files")
        lines.append(f"omitted: {csv(values)}")
    return CommandResult(join_lines(lines))


def _detect_project(root: Path) -> str:
    checks = [
        ("python", "pyproject.toml"),
        ("node", "package.json"),
        ("rust", "Cargo.toml"),
        ("go", "go.mod"),
    ]
    for project, filename in checks:
        if (root / filename).exists():
            return project
    return "unknown"


def _important(root: Path) -> list[str]:
    return [name for name in IMPORTANT_ORDER if (root / name).exists()]


def _ordered(entries: list[str]) -> list[str]:
    rank = {name: index for index, name in enumerate(IMPORTANT_ORDER)}
    return sorted(entries, key=lambda item: (rank.get(item, 99), item))


def _group_sort(item: tuple[str, list[str]]) -> tuple[int, str]:
    return (0 if item[0] == "root" else 1, item[0])

File: src/commands/tree.js
import fs from 'node:fs'; import path from 'node:path';
import { IGNORED_NAMES, TREE_MAX_DEPTH, TREE_MAX_FILES, displayPath } from '../fs.js'; import { csv, joinLines, result } from '../output.js';
const IMPORTANT = ['pyproject.toml','package.json','Cargo.toml','go.mod','README.md','src','tests'];
export function run(root = '.') {
  let files = 0, dirs = 0, truncated = false; const omitted = new Set(), groups = new Map();
  function walk(dir, depth) {
    const rel = path.relative(root, dir), names = fs.readdirSync(dir, { withFileTypes: true }).sort((a,b) => a.name.localeCompare(b.name)); const visible=[]; const children=[];
    for (const entry of names) { if (files >= TREE_MAX_FILES) { truncated=true; break; } if (IGNORED_NAMES.has(entry.name)) { omitted.add(entry.name); continue; } if (entry.isSymbolicLink()) { omitted.add(`${entry.name}@`); continue; } visible.push(entry.name); if (entry.isDirectory()) { if (depth >= TREE_MAX_DEPTH) omitted.add(entry.name); else { dirs++; children.push(path.join(dir,entry.name)); } } else files++; }
    if (visible.length) groups.set(rel || 'root', visible.sort((a,b) => (IMPORTANT.indexOf(a)<0?99:IMPORTANT.indexOf(a))-(IMPORTANT.indexOf(b)<0?99:IMPORTANT.indexOf(b)) || a.localeCompare(b)).slice(0,12));
    for (const child of children) walk(child, depth+1);
  }
  walk(root,0); const project = [['python','pyproject.toml'],['node','package.json'],['rust','Cargo.toml'],['go','go.mod']].find(([,f])=>fs.existsSync(path.join(root,f)))?.[0] ?? 'unknown';
  const lines=[`project=${project} files=${files} dirs=${dirs} important=${csv(IMPORTANT.filter(f=>fs.existsSync(path.join(root,f))))}`];
  for (const [name, entries] of [...groups].sort((a,b)=>a[0]==='root'?-1:b[0]==='root'?1:a[0].localeCompare(b[0]))) lines.push(`${displayPath(name)}: ${csv(entries)}`);
  if (omitted.size || truncated) lines.push(`omitted: ${csv([...omitted].sort().concat(truncated?['max_files']:[]))}`); return result(joinLines(lines));
}
